torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -290,102 +436,114 @@ auto handle_torch_function_no_python_arg_parser(
-            .is(mode_obj),
-        "Defining your mode's `",
-        torch_function_name_str,
-        "` as a classmethod is not supported, please make it a plain method");
-
-    // Blegh.  This accidentally works in PyObject_CallFunctionObjArgs below
-    // because the nullptr terminates the argument list ick ick ick.
-    if (kwargs == nullptr) {
-      ret = py::reinterpret_steal<py::object>(PyObject_CallMethod(
-          mode_obj.ptr(),
-          torch_function_name_str,
-          "OOO",
-          torch_api_function,
-          py_types.ptr(),
-          args));
-    } else {
-      ret = py::reinterpret_steal<py::object>(PyObject_CallMethod(
-          mode_obj.ptr(),
-          torch_function_name_str,
-          "OOOO",
-          torch_api_function,
-          py_types.ptr(),
-          args,
-          kwargs));
-    }
-    if (ret.ptr() == nullptr) {
-      throw python_error();
-    }
-  }
+    // Step 1: Try to dispatch on any user TorchDispatchModes (including infra
+    // modes, which will always be at the bottom of the mode stack).
+    auto ret_ = dispatch_on_mode(
+        args,
+        kwargs,
+        py_types,
+        torch_api_function,
+        is_torch_function,
+        torch_function_name_str);
+    ret = std::get<0>(ret_);
+    mode_obj = std::get<1>(ret_);
+  }
+
+  // Step 2: Try to dispatch based on any user subclasses,
+  // ignoring any subclasses that have a _mode_key field
+  // (corresponding to infra subclasses)
+  // Note: user subclasses should always run *before* infra modes like
+  // proxy/fake. This is handles by having proxy/fake modes return
+  // NotImplemented when they see a user subclass that they don't understand.
   if (ret.ptr() == nullptr || ret.ptr() == Py_NotImplemented) {
-    for (auto& arg : overloaded_args) {
-      py::object torch_function =
-          PyObject_FastGetAttrString(arg, torch_function_name_str);
-      if (!torch_function) {
-        TORCH_INTERNAL_ASSERT(0);
-      }
-
-      // See https://github.com/pytorch/pytorch/issues/63767
-      if (PyObject_FastGetAttrString(torch_function.ptr(), "__self__")
-              .is(py::handle(arg)) &&
-          torch_function.ptr() != torch::disabled_torch_function_impl()) {
-        TORCH_WARN(
-            "Defining your `",
-            torch_function_name_str,
-            "` as a plain method is deprecated ",
-            "and will be an error in future, please define it as a classmethod.");
-      }
-
-      ret = py::reinterpret_steal<py::object>(PyObject_CallFunctionObjArgs(
-          torch_function.ptr(),
-          torch_api_function,
-          py_types.ptr(),
+    auto curr_ret = dispatch_on_subclass(
+        args,
+        kwargs,
+        overloaded_args,
+        py_types,
+        torch_api_function,
+        is_torch_function,
+        torch_function_name_str);
+    if (curr_ret.ptr() != nullptr) {
+      ret = curr_ret;
+    }
+  }
+
+  // Step 3: Try to dispatch based on any infra subclasses,
```

Comment:
Actually, I'm pretty sure this step should be unnecessary too (no need to manually run user subclasses before infra subclasses, because I can rely on our infra subclasses to raise NotImplemented and give user subclasses a chance to run first.)

So in that case, the final dispatch logic ends up being similar (the same) as before.

Right now the dispatching logic is refactored into a separate `dispatch_on_mode` and `dispatch_on_subclass` helper function. If people would prefer, I can try to fully revert to the old logic (mainly since it's less to  review). But otherwise, I'll leave it refactored. Happy to do either.